virtualizarr/zarr.py

```diff
@@ -134,6 +148,61 @@ def replace(
             zarr_format=zarr_format if zarr_format is not None else self.zarr_format,
         )
 
+    def _v3_codec_pipeline(self) -> list:
+        """
+        VirtualiZarr internally uses the `filters`, `compressor`, and `order` attributes
+        from zarr v2, but to create conformant zarr v3 metadata those 3 must be turned into `codecs` objects.
+        Not all codecs are created equal though: https://github.com/zarr-developers/zarr-python/issues/1943
+        An array _must_ declare a single ArrayBytes codec, and 0 or more ArrayArray, BytesBytes codecs.
+        Roughly, this is the mapping:
+        ```
+            filters: Iterable[ArrayArrayCodec] #optional
+            compressor: ArrayBytesCodec #mandatory
+            post_compressor: Iterable[BytesBytesCodec] #optional
+        ```
+        """
+        if self.filters:
+            filter_codecs_configs = [
+                numcodecs.get_codec(filter).get_config() for filter in self.filters
+            ]
+            filters = [
+                dict(name=codec.pop("id"), configuration=codec)
+                for codec in filter_codecs_configs
+            ]
+        else:
+            filters = []
+
+        # Noting here that zarr v3 has very few codecs specificed in the official spec,
+        # and that there are far more codecs in `numcodecs`. We take a gamble and assume
+        # that the codec names and configuration are simply mapped into zarrv3 "configurables".
+        compressor_codec = numcodecs.get_codec(
+            # default to gzip because it is officially specified in the zarr v3 spec
+            dict(id=self.compressor or "gzip")
```

Comment:
Just making sure I understand the meaning of `compressor` here. Is this required to be the compressor the actual array chunks were written with, by whoever wrote the NetCDF / Zarr / etc. file? If so, I don't think the concept of a "default compressor" makes sense here... IIUC, zarr-python will come along looking for some data, follow the ChunkManifest, make a range request for some bytes, and then try to decode them, at which point it would fail (since the data isn't actually gzipped?).

But I see you say that `compressor: ArrayBytesCodec` is mandatory, so we have to have something?

Reply:
>  Is this required to be the compressor the actual array chunks were written with, by whoever wrote the NetCDF / Zarr / etc. file?

Yes. When VirtualiZarr constructs a `ZArray` by reading a valid store, the compressor is read from the store. Even in ZarrV2, [compressors are mandatory](https://zarr-specs.readthedocs.io/en/latest/v2/v2.0.html#metadata), so unless people are reading malformed metadata, this line we're discussing should never fall through to the default. 

The only time a `None` compressor is defaulted into a `gzip` is when someone constructs a `ZArray` in memory and sets `compressor=None`. The unit tests do this. But so does Sean's PR in the `hdf.py` https://github.com/zarr-developers/VirtualiZarr/pull/87/files now I'm thinking VirtualiZarr should throw if someone passes in a `None` compressor to conform to both zarrv2 and zarrv3 semantics.

> compressor: ArrayBytesCodec is mandatory, so we have to have something?

The store we're reading also had to have something, so unless everyone is misinterpreting the spec I think we can assume `compressor` will be set for us. In that case, I think having a default compressor is not necessary and we should probably disallow None compressors.

Reply:
> The only time a None compressor is defaulted into a gzip is when someone constructs a ZArray in memory and sets compressor=None. 

Thanks for that clarification.

> now I'm thinking VirtualiZarr should throw if someone passes in a None compressor to conform to both zarrv2 and zarrv3 semantics.

That sounds right, but probably best as a separate PR.

Reply:
cc @sharkinsspatial, for whether this will affect the compressor stuff in your PR at https://github.com/zarr-developers/VirtualiZarr/pull/87/files. https://github.com/zarr-developers/VirtualiZarr/pull/87/files#diff-d404da2ebd34ec3696e439dfb0ccbb16214cbb9b4f6c5bdcc2083b3eec270f4dR190 specifically I think.

Reply:
TODO: setting default `gzip` is a bad idea, real life Kerchunk stores have compressor=None

Reply:
> we should probably disallow None compressors.

FYI I think https://github.com/zarr-developers/VirtualiZarr/issues/86 represents a legitimate real use case for having no compressor!